src/WinRT.Runtime/IWinRTObject.net5.cs

```diff
@@ -1,188 +1,269 @@
-                    return true;
-                }
-                IObjectReference typedObjRef = (IObjectReference)typeof(IObjectReference).GetMethod("As", Type.EmptyTypes).MakeGenericMethod(vftblType).Invoke(objRef, null);
-                if (!QueryInterfaceCache.TryAdd(interfaceType, typedObjRef))
-                {
-                    typedObjRef.Dispose();
-                }
-                return true;
-            }
-        }
-
-        RuntimeTypeHandle IDynamicInterfaceCastable.GetInterfaceImplementation(RuntimeTypeHandle interfaceType)
-        {
-            var type = Type.GetTypeFromHandle(interfaceType);
-            var helperType = type.GetHelperType();
-            if (helperType.IsInterface)
-                return helperType.TypeHandle;
-            return default;
-        }
-
-        IObjectReference NativeObject { get; }
-        bool HasUnwrappableNativeObject { get; }
-
-        protected ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> QueryInterfaceCache { get; }
-
-        IObjectReference GetObjectReferenceForType(RuntimeTypeHandle type)
-        {
-            return GetObjectReferenceForTypeFallback(type);
-        }
-
-        IObjectReference GetObjectReferenceForTypeFallback(RuntimeTypeHandle type)
-        {
-            if (IsInterfaceImplemented(type, true))
-            {
-                return QueryInterfaceCache[type];
-            }
-            throw new Exception("Interface " + Type.GetTypeFromHandle(type) +" is not implemented.");
-        }
-
-        ConcurrentDictionary<RuntimeTypeHandle, object> AdditionalTypeData { get; }
-
-        object GetOrCreateTypeHelperData(RuntimeTypeHandle type, Func<object> helperDataFactory)
-        {
-            return AdditionalTypeData.GetOrAdd(type, (type) => helperDataFactory());
-        }
-    }
-}
\ No newline at end of file
+#if EMBED
+    internal
+#else
+    public
+#endif
+    interface IWinRTObject : IDynamicInterfaceCastable
+#if NET8_0_OR_GREATER
+        , IUnmanagedVirtualMethodTableProvider
+#endif
+    {
+        bool IDynamicInterfaceCastable.IsInterfaceImplemented(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented)
+        {
+            return IsInterfaceImplementedFallback(interfaceType, throwIfNotImplemented);
+        }
+
+        bool IsInterfaceImplementedFallback(RuntimeTypeHandle interfaceType, bool throwIfNotImplemented)
+        {
+            if (QueryInterfaceCache.ContainsKey(interfaceType))
+            {
+                return true;
+            }
+
+#if NET8_0_OR_GREATER
+            bool vtableLookup = LookupGeneratedVTableInfo(interfaceType, out _, out int qiResult);
+            if (vtableLookup)
+            {
+                return true;
+            }
+            else if (qiResult < 0 && throwIfNotImplemented)
+            {
+                // A qiResult of less than zero means the call to QueryInterface has failed.
+                ExceptionHelpers.ThrowExceptionForHR(qiResult);
+            }

```

Comment:
Any reason we can't use the `QueryInterfaceCache` to remove some of this code?

Reply:
I'm not sure if `QueryInterfaceCache` would help us here, since firstly it's run rather early in the code, and secondly even if I were to add the QI result to `QueryInterfaceCache`, the lack of the interface in `QueryInterfaceCache` wouldn't necessarily mean the QI failed (for instance, if the interface isn't actually a generated COM interface).